neps/nep-0384.md

```diff
@@ -0,0 +1,108 @@
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expectedConnectionEnd))
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+
+// ConnOpenAck relays acceptance of a connection open attempt from chain B back to chain A (this code is executed on chain A).
+function connOpenAck(
+  identifier: Identifier,
+  clientState: ClientState,
+  version: string,
+  counterpartyIdentifier: Identifier,
+  proofTry: CommitmentProof,
+  proofClient: CommitmentProof,
+  proofConsensus: CommitmentProof,
+  proofHeight: Height,
+  consensusHeight: Height) {
+    abortTransactionUnless(consensusHeight < getCurrentHeight())
+    abortTransactionUnless(validateSelfClient(clientState))
+    connection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless((connection.state === INIT && connection.version.indexOf(version) !== -1)
+    expectedConsensusState = getHostConsensusState(consensusHeight) // we need to return a near header at this height.
+    expectedConnectionEnd = ConnectionEnd{TRYOPEN, identifier, getCommitmentPrefix(),
+                             connection.counterpartyClientIdentifier, connection.clientIdentifier,
+                             version, connection.delayPeriodTime, connection.delayPeriodBlocks}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofTry, counterpartyIdentifier, expectedConnectionEnd)) // so we can verify that the counterparty chain has the correct header.
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, connection.counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    connection.state = OPEN
+    connection.version = version
+    connection.counterpartyConnectionIdentifier = counterpartyIdentifier
+    provableStore.set(connectionPath(identifier), connection)
+}
+```
+
+The api `env::block_hash` allows us validate the full header submitted via contract calldata by comparing hashes.
+
+## Rationale and alternatives
+
+A few other designs were considered such as a function `env::block_header` which would return the full header at a given height, but was discouraged against a result of expensive db calls. This mapping of block numbers to headers are cheap enough to be stored on-chain.
+
+
+## Specification
+
+```rust
+/// This function should return a blockhash for 256 of the most reccent blocks
+fn block_hash(number: u32) -> Option<[u8; 32]>;
```

Comment:
```suggestion
/// Returns a blockhash for the `number` block if available.
fn block_hash(number: u32) -> Option<[u8; 32]>;
```

Reply:
If this returns multiple block hashes than this should be `Vec<[u8; 32]>`.

Reply:
This should return a single block hash

Reply:
So `number` is block height?

Reply:
```suggestion
/// This function should return a blockhash for 256 of the most reccent blocks
fn block_hash(number: u64) -> Option<[u8; 32]>;
```

Reply:
+1 - we should call it block_height (as number is ambiguous)